model/riscv_vmem.sail

```diff
@@ -0,0 +1,851 @@
+  PTW_Invalid_PTE   : unit,
+  PTW_No_Permission : unit,
+  PTW_Misaligned    : unit,	     // misaligned superpage
+  PTW_PTE_Update    : unit,	     // PTE update needed but not enabled
+  PTW_Ext_Error	    : ext_ptw_error  // parameterized for errors from extensions
+}
+
+// PRIVATE: only 'to_str' overload is public
+function ptw_error_to_str(e : PTW_Error) -> string = {
+  match e {
+    PTW_Invalid_Addr()	 => "invalid-source-addr",
+    PTW_Access()	 => "mem-access-error",
+    PTW_Invalid_PTE()	 => "invalid-pte",
+    PTW_No_Permission()	 => "no-permission",
+    PTW_Misaligned()	 => "misaligned-superpage",
+    PTW_PTE_Update()	 => "pte-update-needed",
+    PTW_Ext_Error(e)	 => "extension-error"
+  }
+}
+
+// PUBLIC
+overload to_str = {ptw_error_to_str}
+
+// hook for (non-standard) extensions to customize errors reported by page-table
+// walks during address translation; it typically works in conjunction
+// with any customization to check_PTE_permission().
+
+// PRIVATE
+function ext_get_ptw_error(eptwf : ext_ptw_fail) -> PTW_Error =
+  PTW_No_Permission()
+
+// Convert translation/PTW failures into architectural exceptions
+function translationException(a : AccessType(ext_access_type),
+			      f : PTW_Error)
+			     -> ExceptionType = {
+  match (a, f) {
+    (_, PTW_Ext_Error(e))	 => E_Extension(ext_translate_exception(e)),
+    (ReadWrite(_), PTW_Access()) => E_SAMO_Access_Fault(),
+    (ReadWrite(_), _)		 => E_SAMO_Page_Fault(),
+    (Read(_), PTW_Access())	 => E_Load_Access_Fault(),
+    (Read(_), _)		 => E_Load_Page_Fault(),
+    (Write(_), PTW_Access())	 => E_SAMO_Access_Fault(),
+    (Write(_), _)		 => E_SAMO_Page_Fault(),
+    (Execute(), PTW_Access())	 => E_Fetch_Access_Fault(),
+    (Execute(), _)		 => E_Fetch_Page_Fault()
+  }
+}
+
+// PRIVATE
+union PTW_Result = {
+  PTW_Success: (bits(64), bits(64), bits(64), nat, bool, ext_ptw),
+  PTW_Failure: (PTW_Error, ext_ptw)
+}
+
+// ****************************************************************
+// Page Table Walk (PTW)
+
+// Note: 'pt_walk()' is recursive => cannot merge 'val' and 'function' decls
+
+// PRIVATE
+val pt_walk : (SV_Params,
+	       bits(64),		     // virtual addr
+	       AccessType(ext_access_type),  // Read/Write/ReadWWrite/Execute
+	       Privilege,		     // User/Supervisor/Machine
+	       bool,			     // mstatus.MXR
+	       bool,			     // do_sum
+	       bits(64),		     // PT base addr
+	       nat,			     // tree level for this recursive call
+	       bool,			     // global translation,
+	       ext_ptw)			     // ext_ptw
+	      -> PTW_Result
+
+function pt_walk(sv_params,
+		 va,
+		 ac,
+		 priv,
+		 mxr,
+		 do_sum,
+		 pt_base,
+		 level,
+		 global,
+		 ext_ptw) = {
+  let vpn_j	 = vpn_j_of_va(sv_params, va, level);
+  let pte_offset = vpn_j << sv_params.log_pte_size_bytes;
+  let pte_addr	 = pt_base + pte_offset;
+  let pte_phys_addr : xlenbits = pte_addr[(sizeof(xlen) - 1) .. 0];
```

Comment:
Is this right? Physical addresses are 34 bits with SV32.

Maybe a bug in `mem_read_priv()` I guess. Worth adding a TODO comment at least.

Reply:
Yes, Sv32 phys mem addrs are 34 bits, not 32 bits.  But it is used in 'mem_read_priv()' [riscv_mem.sail, L180], where the address argument is declared as 'xlenbits', so until mem_read_priv() is changed, we have to use 32 bits.

I don't know if it's straightforward to fix 'mem_read_priv()'.  As I understand it, in RV32, phys mem addrs are 32(34) bits when Sv32 is not (is) supported.  Even if Sv32 is supported, what happens when we run with M privilege (are phys addrs 32 bits? Or extended to 34 bits)?

I've mentioned this several times in our meetings (perhaps I should add an "issue" so it doesn't slip thru the cracks), and I'll also add a local TODO comment here in the vmem code.

Reply:
Physical addresses at 34 bits if Sv32 is implemented, but Mmode cannot access beyond 32 bits 
**unless** it sets (mstatus.MPRV==1) && (mstatus.MPP !=mmode) && (satp.mode!=0).
If those are set correctly, then it will use VM translation and use all 34 bits. 
Otherwise, the upper 2 bits can't be used in Mmode, and they're set to zero.